src/core/geometry/qgsgeos.cpp

```diff
@@ -1440,18 +1407,38 @@ GEOSCoordSequence* QgsGeos::createCoordinateSequence( const QgsCurveV2* curve )
   try
   {
     coordSeq = GEOSCoordSeq_create_r( geosinit.ctxt, numPoints, coordDims );
-    for ( int i = 0; i < numPoints; ++i )
+    if ( precision > 0. )
     {
-      QgsPointV2 pt = line->pointN( i ); //todo: create method to get const point reference
-      GEOSCoordSeq_setX_r( geosinit.ctxt, coordSeq, i, pt.x() );
-      GEOSCoordSeq_setY_r( geosinit.ctxt, coordSeq, i, pt.y() );
-      if ( hasZ )
+      for ( int i = 0; i < numPoints; ++i )
       {
-        GEOSCoordSeq_setOrdinate_r( geosinit.ctxt, coordSeq, i, 2, pt.z() );
+        QgsPointV2 pt = line->pointN( i ); //todo: create method to get const point reference
+        GEOSCoordSeq_setX_r( geosinit.ctxt, coordSeq, i, std::round( pt.x() / precision ) * precision );
+        GEOSCoordSeq_setY_r( geosinit.ctxt, coordSeq, i, std::round( pt.y() / precision ) * precision );
```

Comment:
And in case one wonders: the usage of std::round instead of qRound is a conscious decision, because std::round returns a double, whereas qRound returns an int, which can easily overflow.